### PR TITLE
BOLT 7,11: Added an optional `min_final_cltv_expiry` field in BOLT 11

### PR DESCRIPTION
**Pull request by pm47**
This PR competes with #252.

It clarifies how `channel_update` messages are only to be used in the context of relaying payments, and how both htlc amounts and expiries are to be calculated backwards from the values provided in the payment request.

This means that we don't need the `channel_update` for the first channel in a route, which allows making a payment through a channel which hasn't had any announcements yet.

Note that the clever calculation leading to the default value of 9 for `min_final_cltv_expiry`
has been made in #253.

### CODE REVIEW

File: 07-routing-gossip.md [hunk 1]
```diff
@@ -422,10 +433,6 @@ Notice that this is a node policy and MUST NOT be enforced by peers, e.g., by cl
 
 ## Recommendations for Routing
 
-As the fee is proportional, it must be calculated backwards from the
-destination to the source: only the amount required at the final
```
Comment (Roasbeef) [inline 07-routing-gossip.md]: Why was this removed?

File: 11-payment-encoding.md [hunk 2]
```diff
@@ -162,6 +166,11 @@ A reader MUST skip over unknown fields, an `f` field with unknown
 A reader MUST check that the SHA-2 256 in the `h` field exactly
 matches the hashed description.
 
+A reader MUST use a greater value for the `cltv_expiry` of the last
```
Comment (Roasbeef) [inline 11-payment-encoding.md]: Feels like this section should instead be placed somewhere in #253 as this section currently defines what implementers should note when _parsing_.

File: 07-routing-gossip.md (hunk 1 above)
Comment (Roasbeef) [reply 07-routing-gossip.md]: Ah ok I see it was moved above!

File: 11-payment-encoding.md [hunk 3]
```diff
@@ -109,6 +109,7 @@ Currently defined Tagged Fields are:
 * `n` (19): `data_length` 53.  The 33-byte public key of the payee node.
 * `h` (23): `data_length` 52.  256-bit description of purpose of payment (SHA256).  This is used to commit to an associated description which is too long to fit, such as may be contained in a web page.
 * `x` (6): `data_length` variable.  `expiry` time in seconds (big-endian). Default is 3600 (1 hour) if not specified.
+* `c` (24): `data_length` (16 bits, big-endian). Minimum `cltv_expiry` to use for the last htlc. Default is 9 if not specified.
```
Comment (Roasbeef) [inline 11-payment-encoding.md]: I think the `data_length` section here should be modified to be 4 (4 * 5 = 20, so 4 unused bits).

Comment (rustyrussell) [reply 11-payment-encoding.md]: Make it variable, like x.  If you can fit it in 5 bits, great.  We should enforce a maximum of 6 (ie. fits in 32 bits) though, that's a separate PR.

Comment (rustyrussell) [reply 11-payment-encoding.md]: Can we rename it to `final_cltv` or `min_final_cltv` because we already have a `cltv_expiry` field in the onion...

Comment (jimpo) [inline 11-payment-encoding.md]: Should the data length be specified as a multiple of 5 bits? 10 bits seems like plenty.

Comment (Roasbeef) [reply 11-payment-encoding.md]: In the `update_channel` messages the delta is expressed using 16 bits. Seem like we should maintain uniformity of the allotted size of "delta" values throughput the spec.

Agreed on using a distinct name.

Comment (rustyrussell) [reply 11-payment-encoding.md]: Ack...

File: 07-routing-gossip.md (hunk 1 above)
Comment (pm47) [reply 07-routing-gossip.md]: Yes I thought it made more sense there and it wasn't really a recommendation anyway.

File: 11-payment-encoding.md (hunk 3 above)
Comment (pm47) [reply 11-payment-encoding.md]: Yes, I chose 16 bits to be consistent with `update_channel` as laolu points out. Will update the name.

File: 11-payment-encoding.md (hunk 2 above)
Comment (pm47) [reply 11-payment-encoding.md]: Agreed, I removed it

Comment (pm47) [reply 11-payment-encoding.md]: I also added a section in the rationale and took inspiration in [this comment of yours](https://github.com/lightningnetwork/lightning-rfc/pull/252#issuecomment-335260301)

**Comment by rustyrussell** [comment]
Please change to `data_length` variable like the `x` field above.

And change:

    A writer MAY include one `x` field, which SHOULD use the minimum `data_length`  possible.

To:
    A writer MAY include one `x` field, and MAY include one `c` field.  It SHOULD use  the minimum `data_length`  possible for these fields.

I will add a separate PR which lets the reader reject x and c fields > 60 bits (handling things > 64 bits is completely unnecessary complexity, since they're supposed to be minimum length and 60 bits for either is insane).

**Comment by pm47** [comment]
Well I explained why I didn't make this change, it is because it wouldn't be consistent with how the expiry is handled in the `r` field?

**Comment by rustyrussell** [comment]
Hmm, sorry, I must have missed that :(

In the 'r' field we have other values which are already 8 bit aligned.  I guess this only saves 2 bytes, and I'm not sure about whether `c` will be common; depends on how well 9 works out?  But `x` is already variable, so the logic is trivial...

**Comment by pm47** [comment]
I don't really mind. Let's see what @Roasbeef says? He seemed more in favor of consistency.

**Comment by Roasbeef** [comment]
I'm in favor of the current state, so allowing `x` to be variable size as well (ensuring it's minimally packed). This lets us re-use the existing code already used for the expiry value.

**Comment by Roasbeef** [comment]
Ah, actually as the expiry delta (an values with different names but similar purposes across the spec) are _always_ 16-bits, I'd argue that it should be consistent and use a _fixed_ length 16-bit value. 

FWIW, for now `lnd` will _always_ set it's own `c` value 😉

**Comment by rustyrussell** [comment]
Well, the default value c=9 was basically chosen because that's what  ECLAIR did and it seems reasonable.  If people want a different default, please speak up now! @Roasbeef @pm47

**Comment by pm47** [comment]
Default c=9 is fine with me!

**Comment by rustyrussell** [comment]
Assigned to @pm47 for him to decide whether to commit or revert c varlen patch before merge.

**Comment by pm47** [comment]
I'm keeping it!

**Comment by rustyrussell** [comment]
Soo... I think we have consensus?  If so, please apply, so I can apply #253 (or prompt @Roasbeef for final review...)

**Comment by pm47** [comment]
@sstone told me there wasn't any objection during monday's call so I merged as per our guidelines. 

@Roasbeef please let me know if I misunderstood.

File: 07-routing-gossip.md (hunk 1 above)
Comment (zinmar222269-cloud) [reply 07-routing-gossip.md:426]: aed3d28d5a202a0dd7b037fca4c04f58242fa909
